### PR TITLE
4th rework - updated mruudecode.c to handle uuencoded attachments

### PR DESCRIPTION
**Pull request by csb0730**
This is a new pull request for uuencoded data handling.
Base of csb0730/deltachat-core is current state.
Only _mruudecode.c_ is updated to add all needed functions.

Changes are based on #160 PR

:-)

### CODE REVIEW

**Comment by r10s** [comment]
looks good :) the next few days i have probably no time for testing, but after that :)

File: src/mruudecode.c [hunk 1]
```diff
@@ -59,8 +132,727 @@ char* mruudecode_do(const char* text, char** ret_binary, size_t* ret_binary_byte
...
+    uu_part_loop_cnt = 0;
+
 	return NULL;
 }
 
 
+
+
+/**
+ *  From here on new code from  Christian Schneider <schneider17@gmx.de>  follows
```
Comment (r10s) [inline src/mruudecode.c]: i think this header is superfluous now.

**Comment by r10s** [comment]
made some minor changes, btw. Delta Chat uses tabs instead of spaces.

**Comment by r10s** [comment]
(didn't realize that i can also push to your fork :)

File: src/mruudecode.c [hunk 2]
```diff
@@ -19,10 +19,83 @@
...
+#define CRLF "\r\n" //1310 dez
+#define CR   "\r"   //13
+#define LF   "\n"   //10
+
+// prevent infinite loops
+#define MAX_DECODING_LOOPS 20
+
+
+// prevent infinite loops 
+static int      uu_part_loop_cnt    = 0;
```
[the remark this answers is not in the document]
Comment (csb0730) [reply src/mruudecode.c]: I just made this global to ease parameter handover ;-)

File: src/mruudecode.c [hunk 3]
```diff
@@ -19,10 +19,83 @@
...
+
+// keeps line end pattern for all functions
+static char*    line_end_pattern    = NULL;
```
[the remark this answers is not in the document]
Comment (csb0730) [reply src/mruudecode.c]: I just took out handover line_end_pattern to any function to ease parameter hand over. My first implementation used no globals ;-)
Yes the only method to ease this is to unify line ends.
If this unifying is done before the job here this simplifies the complete job here and all occurrences of working with line ends will get much easier. But You need to copy the buffer
one time (the only disadvantage).

But isn't there a standard which demands CRLF as an email standard?

Ok, I'll check and come back.

File: src/mruudecode.c [hunk 4]
```diff
@@ -59,8 +132,727 @@ char* mruudecode_do(const char* text, char** ret_binary, size_t* ret_binary_byte
...
+char* mr_find_uuencoded_part (const char* msgtxt){
+    /**
+     *  Find uuencoded part in msgtxt & return it's position.
+     *  If no part is found NULL is returned.
+     */
+     
+    char* uu_part_pos = NULL;
+    int len_line_end_pattern = strlen(line_end_pattern);
+    
+    char tofind[20] = "";
```
[the remark this answers is not in the document]
Comment (csb0730) [reply src/mruudecode.c]: Looks more easy :)

But to secure uuencode detection the second space after the number "begin 666 filename" should be checked too.

**Comment by csb0730** [comment]
Hi Björn, just saw that You started with that PR :)

File: src/mruudecode.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (csb0730) [reply src/mruudecode.c]: mr_unify_lineends():

Is it really removing CRs or is it replacing by space?

CRLF -->   "LF"   - 2 chars to 1 char
CRLF -->   "_LF" - 2 chars to 2 chars  ( _ = space)

?

**Comment by csb0730** [comment]
Ok to summarize requested changes I see now following steps necessary:

1. unify line ends. This will make every handling much easier.
    mr_detect_line_end() not necessary. Instead use mr_unify_lineends() before any handling.
2. don't use globals
3. don't use a loop limiter internally.
4. simplify all functions for simple LF line end handling.

That's it?

**Comment by r10s** [comment]
> That's it?

yes, roughly :)

File: src/mruudecode.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (r10s) [reply src/mruudecode.c]: > But to secure uuencode detection the second space after the number "begin 666 filename" should be checked too.

yes - and we can also check if the number is valid, if the following lines are formatted etc. if there is an error, we cleanup and stop - meanwhile Delta Chat has no problems with large texts, so this is fine.

i just wanted to have a _rough check_  that says "there might be a uuencoded part, it's worth checking more exactly".
this _rough check_ should not allocate memory, and should be fast.
for the the subsequent _exact check_ we can then copy the buffer, unify lineends etc.

[the remark this answers is not in the document]
Comment (csb0730) [reply src/mruudecode.c]: Hi @r10s, I'm not really in line with your code above!

Whether the rough check needs to be done before calling mruudecode_do()

      if( rough check ){
          workbuf = strdup(buf);
          mr_unify_lineends(buf);
          while( mruudecode_do(buf, ...) ){
              // loop to handle all parts
              // and hands over a unified message to the rest of the code !!  Q: is this ok???
          }
      }

or all code including unifying into mruudecode_do(buf, ...)

But even then: a unified version of the message is returned!

But the code snipped when first called mruudecode_do() and then rough check cannot work.

**Comment by csb0730** [comment]
Hi @r10s, please see my comment [here](https://github.com/deltachat/deltachat-core/pull/161#discussion_r191276974)

Meanwhile I separated the test functions to a separate local file and cleaned mruudecode.c.
I'm starting now to with point 4. to clean all functions only to work with simple LF as line separator
but with same approach I did.

But see my comments in the code snipped: What about the returned remaining text?
Is it ok to have it only unified to LF? If not we have to deal with mixed line ends again.

It would be bad to handle a mixed approach.

**Comment by csb0730** [comment]
@r10s, done so far, waiting for comments on last questions.

**Comment by csb0730** [comment]
May change the file to new naming conventions?

**Comment by r10s** [comment]
> May change the file to new naming conventions?

yes, this would be good. also the referenced functions can be renamed to dc_*() instead of mr()

**Comment by csb0730** [comment]
ok, will start and let you know when done. will be glad when really include it :-)

nice evening :)

Am 5. Juli 2018 10:44:43 MESZ schrieb "Björn Petersen" <notifications@github.com>:
>> May change the file to new naming conventions?
>
>yes, this would be good. also the referenced functions can be renamed
>to dc_*() instead of mr()
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/deltachat/deltachat-core/pull/161#issuecomment-402650143

-- 
Diese Nachricht wurde von meinem Android-Gerät mit K-9 Mail gesendet.

**Comment by csb0730** [comment]
Will be replaced by 5th rework (renaming all functions from mr... to DC_...)

**Pull request closed without merging**